hudi-utilities/src/main/java/org/apache/hudi/utilities/schema/JsonSchemaToAvroSchemaConverter.java

```diff
@@ -0,0 +1,175 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.utilities.schema;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class JsonSchemaToAvroSchemaConverter {
+  public static final String ONE_OF = "oneOf";
+  public static final String TYPE = "type";
+  public static final String NAME = "name";
+  public static final String PROPERTIES = "properties";
+  public static final String FIELDS = "fields";
+  public static final String RECORD = "record";
+  public static final String MARQETA_JCARD_NAMESPACE = "com.marqeta.jcard";
+
+  /**
+   * Handles the oneOf special datatype. Returns an array of [ null, object ]
+   * @param jsonArray
+   * @param name
+   * @return
+   * @throws Exception
+   */
+  private static JsonArray handleOneOf(JsonArray jsonArray, String name) throws Exception {
+    JsonArray types = new JsonArray();
+    for (JsonElement oneOfs: jsonArray) {
+      String oneOfType = oneOfs.getAsJsonObject().get(TYPE).getAsString();
+      // if  type is not null it's a json object
+      if (oneOfType.equalsIgnoreCase("null")) {
+        // handle null
+        types.add("null");
+      } else {
+        Object resp = jsonTypeToAvroType((JsonObject) oneOfs);
+        if (resp instanceof String) {
+          // handle simple datatype
+          types.add((String) resp);
+        } else {
+          // handle object
+          JsonObject avroObject = new JsonObject();
+          avroObject.addProperty(NAME, name);
+          avroObject.addProperty(TYPE, RECORD);
+          avroObject.add(FIELDS, (JsonArray) resp);
+          types.add(avroObject);
+        }
+      }
+    }
+    return types;
+  }
+
+  /**
+   * Maps the json data types to avro data types. Does not handle array as of now
+   * @param property json datatype
+   * @return Object avro datatype
+   * @throws Exception
+   */
+  private static Object jsonTypeToAvroType(JsonObject property) throws Exception {
+    String jsonType = property.get("type").getAsString();
+
+    switch (jsonType) {
+      case "integer":
```

Comment:
Aren't there some static constants / type enums defined in gson for these? If not, it would be better to define static constants in our code.

Reply:
yes, we could. This PR is not ready for review yet. I am still working on using the `JsonSchemaDeserializer` when deserializing events read from the kafka source. Looking at the `JsonKafkaSource` class, everything seems to be converted into a String first and then `avro` later. That becomes a problem when reading decimal numbers as the string conversion leaves the value as `AA==` for a decimal value of `0.00`. Now when the value is converted back to `AVRO` it throws a wrong datatype exception as it was expecting a `double` and not a `string`.

Reply:
I am planning to introduce a `JsonSchemaKafka` source that reads the consumer records and keeps the type information instead of converting them to string, before conversion to Avro. Let me know if you see a better way to do this